In [12]:
import numpy as np

def riemann_derecha_x3(b, n):
    dx = b/n
    i = np.arange(1, n+1)   # derecha: 1..n
    x = i*dx
    return np.sum(x**3)*dx

b = 2.0
n = 3

R3 = riemann_derecha_x3(b, n)

print(f"Suma de Riemann (n={n}) =", R3)
print("Valor exacto =", b**4/4)

# Error
error_abs = abs(b**4/4 - R3)
error_rel = error_abs / (b**4/4)

print("Error absoluto =", error_abs)
print("Error relativo =", error_rel)
print("Error relativo (%) =", error_rel*100)

Suma de Riemann (n=3) = 7.111111111111111
Valor exacto = 4.0
Error absoluto = 3.1111111111111107
Error relativo = 0.7777777777777777
Error relativo (%) = 77.77777777777777


<div style="background-color:#ffe6f0; padding:18px; border-radius:12px; border:1px solid #ffb3d1; color:#1a1a1a;">

🌸 <span style="color:#c2185b; font-size:22px;"><b>Ejercicios de entrega</b></span>

**Problema:** Hallar el área exacta de la región entre la curva $y=x^3$ y el eje $x$ en $[0,b]$, usando **sumas de Riemann a la derecha** con subintervalos de igual ancho.

---

**1) Partición del intervalo (con $b=2$ y $n=3$)**

$$
\Delta x=\frac{b}{n}=\frac{2}{3}
$$

Puntos:
$$
x_i=\frac{ib}{n},\quad i=0,1,2,3
$$

Extremos derechos:
$$
x_1=\frac{2}{3},\quad x_2=\frac{4}{3},\quad x_3=2
$$

---

**2) Suma de Riemann (derecha)**

$$
R_3=\sum_{i=1}^{3} f(x_i)\,\Delta x
$$

Con $f(x)=x^3$:

$$
R_3=\left[\left(\frac{2}{3}\right)^3+\left(\frac{4}{3}\right)^3+(2)^3\right]\left(\frac{2}{3}\right)
$$

$$
R_3=\left[\frac{8}{27}+\frac{64}{27}+8\right]\frac{2}{3}
=\left[\frac{8}{3}+8\right]\frac{2}{3}
=\frac{64}{9}\approx 7.1111
$$

---

**3) Área exacta**

$$
A=\frac{b^4}{4}=\frac{2^4}{4}=4
$$

---

**4) Errores (comparando con el valor exacto $A=4$)**

- Error absoluto:
$$
|A-R_3|=|4-7.1111|\approx 3.1111
$$

- Error relativo:
$$
\frac{|A-R_3|}{|A|}\approx \frac{3.1111}{4}\approx 0.7778
$$

- Error relativo porcentual:
$$
\approx 77.7778\%
$$

---

**Conclusión**

Como $x^3$ es creciente en $[0,2]$, la suma de Riemann a la derecha **sobreestima** el área. Con $n=3$ la sobreestimación es grande, pero al aumentar $n$ la aproximación mejora y converge al valor real.

</div>

<div style="background-color:#ffe6f0; padding:18px; border-radius:12px; border:1px solid #ffb3d1; color:#1a1a1a;">

🌸 <span style="color:#c2185b; font-size:22px;"><b>Ejercicio para entregar — Trapecio compuesto (cohete)</b></span>

Se estima la distancia vertical recorrida por un cohete entre $t=8$ y $t=30$ s:

$$
x=\int_{8}^{30}\left(2000\ln\left(\frac{140000}{140000-2100t}\right)-9.8t\right)\,dt
$$

Defino la función:

$$
f(t)=2000\ln\left(\frac{140000}{140000-2100t}\right)-9.8t
$$

---

**a) Estimación con la regla del trapecio compuesto ($n=3$)**

Divido el intervalo en 3 subintervalos iguales:

$$
h=\frac{b-a}{n}=\frac{30-8}{3}=7.3333
$$

Nodos:

$$
t_0=8,\quad t_1=15.3333,\quad t_2=22.6667,\quad t_3=30
$$

Evaluaciones:

$$
f(t_0)\approx 177.2667,\quad f(t_1)\approx 372.4629,\quad f(t_2)\approx 608.8976,\quad f(t_3)\approx 901.6740
$$

Fórmula del trapecio compuesto:

$$
T_3=\frac{h}{2}\,[f(t_0)+2f(t_1)+2f(t_2)+f(t_3)]
$$

$$
\boxed{T_3 \approx 11152.7591}
$$

---

**b) Error absoluto**

Tomo como valor real el resultado numérico de la integral:

$$
I \approx 11061.3355
$$

$$
E_a = |I - T_3|
$$

$$
\boxed{E_a \approx 91.4236}
$$

---

**c) Error relativo**

$$
E_r = \frac{|I - T_3|}{|I|}
$$

$$
\boxed{E_r \approx 0.008265 \;\;(\approx 0.8265\%)}
$$

---

**Conslusión**

Con solo $n=3$, el método del trapecio compuesto ya da una aproximación bastante cercana al valor real (error menor al 1%).  
A diferencia del trapecio simple, aquí se consideran puntos intermedios, lo que mejora notablemente la precisión.

</div>

In [14]:
import numpy as np
from scipy.integrate import quad

def f(t):
    return 2000*np.log(140000/(140000 - 2100*t)) - 9.8*t

# Intervalo
a = 8.0
b = 30.0

n = 3
h = (b - a) / n


x = np.linspace(a, b, n + 1)   # x0..x3
y = f(x)

# Trapecio compuesto
T3 = (h/2) * (y[0] + 2*np.sum(y[1:-1]) + y[-1])

# Valor "real" (numérico)
I, _ = quad(f, a, b)

# Errores
error_abs = abs(I - T3)
error_rel = error_abs / abs(I)

# Imprimir resultados
print(f"n = {n}")
print("h =", h)
print("nodos x =", x)
print("valores f(x) =", y)

print("\nTrapecio compuesto (T3) =", T3)
print("Valor real (I) =", I)

print("\nError absoluto =", error_abs)
print("Error relativo =", error_rel)
print("Error relativo (%) =", error_rel * 100)

n = 3
h = 7.333333333333333
nodos x = [ 8.         15.33333333 22.66666667 30.        ]
valores f(x) = [177.26674302 372.4628616  608.89755459 901.67400151]

Trapecio compuesto (T3) = 11152.759115356115
Valor real (I) = 11061.335535080992

Error absoluto = 91.42358027512273
Error relativo = 0.008265148452026713
Error relativo (%) = 0.8265148452026713


In [11]:
import numpy as np
import pandas as pd
from scipy.integrate import quad
from IPython.display import display

def f(t):
    return 2000*np.log(140000/(140000 - 2100*t)) - 9.8*t

a, b = 8.0, 30.0

def trapecio_compuesto(f, a, b, n):
    h = (b - a) / n
    x = np.linspace(a, b, n + 1)
    y = f(x)
    T = (h/2) * (y[0] + 2*np.sum(y[1:-1]) + y[-1])
    return h, x, y, T

# Valor real 
I, _ = quad(f, a, b)
print("===== Valor real =====")
print(f"I = {I:.6f}")

n = 3
h, x, y, T = trapecio_compuesto(f, a, b, n)

err_abs = abs(I - T)
err_rel = err_abs / abs(I)

print(f"\n===== Trapecio compuesto (n={n}) =====")
print("h =", h)
print("nodos =", x)
print("f(x) =", y)

print(f"\nT = {T:.6f}")
print(f"Error absoluto = {err_abs:.6f}")
print(f"Error relativo = {err_rel:.6f}")
print(f"Error relativo (%) = {err_rel*100:.3f}%")


filas = []
for n in range(1, 9):
    _, _, _, Tn = trapecio_compuesto(f, a, b, n)
    ea = abs(I - Tn)
    er = ea / abs(I)
    filas.append([n, Tn, ea, er, er*100])

tabla = pd.DataFrame(filas, columns=[
    "n",
    "x (trapecio compuesto)",
    "error absoluto",
    "error relativo",
    "error relativo (%)"
])

display(
    tabla.style
    .background_gradient(cmap="RdPu")
    .format({
        "x (trapecio compuesto)": "{:.4f}",
        "error absoluto": "{:.4f}",
        "error relativo": "{:.5f}",
        "error relativo (%)": "{:.3f}"
    })
    .set_properties(**{'text-align': 'center'})
    .set_table_styles([
        {'selector': 'th',
         'props': [
             ('background-color', '#FFC0CB'),
             ('color', 'black'),
             ('font-weight', 'bold'),
             ('text-align', 'center')
         ]}
    ])
)

===== Valor real =====
I = 11061.335535

===== Trapecio compuesto (n=3) =====
h = 7.333333333333333
nodos = [ 8.         15.33333333 22.66666667 30.        ]
f(x) = [177.26674302 372.4628616  608.89755459 901.67400151]

T = 11152.759115
Error absoluto = 91.423580
Error relativo = 0.008265
Error relativo (%) = 0.827%


<div style="background-color:#ffe6f0; padding:18px; border-radius:12px; border:1px solid #ffb3d1; color:#1a1a1a;">

🌸 <span style="color:#c2185b; font-size:22px;"><b>Ejercicio para entregar — Método del trapecio compuesto(cohete)</b></span>

En este ejercicio se estima la distancia recorrida por un cohete entre $t = 8$ y $t = 30$ segundos usando el método del trapecio compuesto.

La función a integrar es:

$$
f(t) = 2000 \ln\left(\frac{140000}{140000 - 2100t}\right) - 9.8t
$$

---

**Aproximación con $n = 3$**

Se divide el intervalo en 3 subintervalos iguales:

$$
h = \frac{b - a}{n} = \frac{30 - 8}{3}
$$

Se evalúa la función en los nodos y se aplica la fórmula del trapecio compuesto.

El valor aproximado obtenido fue:

$$
T_3 \approx 11152.76
$$

---

**Cálculo del error**

Para evaluar la precisión, se compara con el valor real (calculado numéricamente):

$$
I \approx 11061.34
$$

- Error absoluto:
$$
|I - T_3| \approx 91.42
$$

- Error relativo:
$$
\approx 0.00827
$$

- Error relativo porcentual:
$$
\approx 0.827\%
$$

---
**Conclusión**

Se observa que el método del trapecio compuesto da una buena aproximación incluso con pocos subintervalos.

Además, al aumentar el número de particiones $n$, el error disminuye, lo que muestra la convergencia del método hacia el valor real.